### PR TITLE
Add data frame as new input format

### PR DESCRIPTION
**Pull request by Sukikui**
Hi,

I'm submitting this pull request to suggest a change in the way you handle data entry, I'm a member of a biomedical research team and your package is indispensable to us. Currently, we read data from a CSV file (or other) and then convert it to Pandas data frame. I propose that we start using Pandas data frames as a new format for reading data directly, without the need for an intermediate CSV file.

This change would simplify our code and make it more efficient, as we could take advantage of pandas features to read and manipulate data and use the fitter in a loop. I tested my features to make sure it works as intended. I let you review my code and understand how work the features in the [README](https://github.com/Sukikui/DFImpedanceFitter). So make attention to not merge my README 🥲

Let me know what you think, and I'll be happy to make further changes if necessary.

Thanks!

### CODE REVIEW

**Comment by j-zimmermann** [comment]
Hi,
Thanks a lot for your contribution and your suggestion. It's great to hear that you actively use ImpedanceFitter :)
I'll check the PR this weekend and comment in greater detail / merge it then.

Cheers,
Julius

**Comment by Sukikui** [comment]
Ok fine :)

File: README.md
Comment (j-zimmermann) [inline README.md:1]: Please move this description to the utils file and the `available_file_formats` function:
https://github.com/j-zimmermann/ImpedanceFitter/blob/853699832711054d032cfa52e8ab3d6b2ec179a3/impedancefitter/utils.py#L533

Then, it will appear in the documentation as here:
https://impedancefitter.readthedocs.io/en/latest/utils.html#impedancefitter.utils.available_file_format

The original README should be preserved.

File: impedancefitter/fitter.py [hunk 1]
```diff
@@ -134,10 +135,10 @@ class Fitter(object):
         Exists only when :meth:`run` was called.
     """
 
-    def __init__(self, inputformat, directory=None, **kwargs):
+    def __init__(self, inputformat, directory=None, df=None, **kwargs):
         """ initializes the Fitter object
         """
-
+        self.df = df
```
Comment (j-zimmermann) [inline impedancefitter/fitter.py:141]: I would not add the dataframe to the attributes of the Fitter object.
Just use it in the entire function.

File: impedancefitter/fitter.py [hunk 2]
```diff
@@ -134,10 +135,10 @@ class Fitter(object):
         Exists only when :meth:`run` was called.
     """
 
-    def __init__(self, inputformat, directory=None, **kwargs):
+    def __init__(self, inputformat, directory=None, df=None, **kwargs):
```
Comment (j-zimmermann) [inline impedancefitter/fitter.py:138]: I would suggest to use only `DF` as the new inputformat.
Then, add the information that you had in the `inputformat` string as kwargs:
Instead of `DF Freq-Re1-Im1` use `inputformat=DF`, `df_freq_column=Freq`, `df_re_column=Re`, `df_im_column=Im`. The limitation for the user would be to make sure that the columns have the same pattern and are numbered with a clear indexing (e.g., 1-indexed as in your example).

Would that suit your purpose?

File: impedancefitter/fitter.py [hunk 3]
```diff
@@ -153,16 +154,19 @@ def __init__(self, inputformat, directory=None, **kwargs):
         if self.fileList is None:
             self.fileList = sorted(os.listdir(self.directory))
         read_data_sets = 0
+
         for filename in self.fileList:
```
Comment (j-zimmermann) [inline impedancefitter/fitter.py:158]: Here is kind of a broken logic now: with the dataframe, there is no need to go through a file list.
I would separate this.
If you do not have an idea, let me know and I will do it.

File: impedancefitter/readin.py [hunk 4]
```diff
@@ -23,6 +23,65 @@
 
 logger = logging.getLogger(__name__)
 
+def _readin_Data_from_formated_dataframe(formated_df, minimumFrequency=None, maximumFrequency=None):
+    # sort frequencies to be on the safe side
+    tmp = formated_df.values
```
Comment (j-zimmermann) [inline impedancefitter/readin.py:28]: Why do you extract the values? 
It would be much easier to just use the columns, right?
If you add two keywords that identify the right columns (frequency, real and imaginary part) that should be possible, right?

File: impedancefitter/readin.py [hunk 5]
```diff
@@ -23,6 +23,65 @@
...
+    omega = 2. * np.pi * f
+    # construct complex-valued array from float data
+    zarray = np.zeros((np.int32((values.shape[1] - 1) / 2), values.shape[0]), dtype=np.complex128)
+
+    for i in range(np.int32((values.shape[1] - 1) / 2)):  # will always be an int(always real and imag part)
+        zarray[i] = values[:, (i * 2) + 1] + 1j * values[:, (i * 2) + 2]
+    return omega, zarray
+
+
+def readin_Data_from_dataframe(df, format, minimumFrequency=None, maximumFrequency=None):
```
Comment (j-zimmermann) [inline impedancefitter/readin.py:60]: I think this function is enough.
Just take the dataframe, sort it by frequency, filter all frequencies that are less / greater than minimum / maximum frequency and then combine real and imaginary part to a complex-valued array.
Or am I missing something?

File: impedancefitter/readin.py [hunk 6]
```diff
@@ -66,38 +125,7 @@ def readin_Data_from_collection(filepath, fileformat, delimiter=None,
         EIS = pd.read_csv(filepath, delimiter=delimiter, header=header)
     else:
         raise NotImplementedError("File type not known")
-
-    # sort frequencies to be on the safe side
```
Comment (j-zimmermann) [inline impedancefitter/readin.py:70]: Why did you delete this part?

**Review by j-zimmermann — COMMENTED** [review]
I checked your contributions in detail.
I very much like the idea to use DataFrames as they have many advantages.
However, the project was written with the idea in mind to support various data formats as produced by the different devices.
That is currently the major drawback because the whole internal code structure is based on that and would be changed now.
My idea would be: the filename related part (reading in, storing names, etc.) and the data processing need to be decoupled.
Have a look if my comments make sense.
If I missed anything, please let me know. Otherwise I will refactor the code accordingly.
The data frame

**Comment by Sukikui** [comment]
Hi, I think I've understood your expectations, I've separated all the methods in `fitter.py` and `readin.py`. I've also put the `df` parameter in the **kwargs along with `df_freq_column`, `df_real_column` and `df_imag_column`.

So I let your review this and refactor something you think it's necessary. Also I've handled exceptions related to dataframe parameters in **kwargs in `fitter.py` (I check if they are None). But if you think it's more coherent to put them in `readin.py` I'll let you move them or let me know so I can fix it quickly! Thanks for your review :)

File: impedancefitter/utils.py
Comment (j-zimmermann) [inline impedancefitter/utils.py:1]: Thanks, this looks good!

File: impedancefitter/readin.py
Comment (j-zimmermann) [inline impedancefitter/readin.py:1]: Also this change looks good, thank you!

**Review by j-zimmermann — COMMENTED** [review]
Thanks for your changes!

I still don't like the different treatment between DataFrame and FileList but I would refactor this after merging and ask you to review it.

To finish the merge: could you please add a test to https://github.com/j-zimmermann/ImpedanceFitter/blob/master/impedancefitter/tests/test_main.py and to https://github.com/j-zimmermann/ImpedanceFitter/blob/master/impedancefitter/tests/test_readin.py where you test the new functionality.
That would be very helpful so that it remains functional also in future versions.

**Comment by Sukikui** [comment]
I added the dataframe test to `test_main.py` and `test_readin.py`. Everything's ok, but I'll let you confirm that my commit to `test_main.py` is what you expected because you are testing the CSV format in the `fitter()` and `fitter2()` test functions, so I've basically just added `fitter_df()` and `fitter2_df()`.

**Comment by Sukikui** [comment]
> Thanks for your changes!
> 
> I still don't like the different treatment between DataFrame and FileList but I would refactor this after merging and ask you to review it.
> 
> To finish the merge: could you please add a test to https://github.com/j-zimmermann/ImpedanceFitter/blob/master/impedancefitter/tests/test_main.py and to https://github.com/j-zimmermann/ImpedanceFitter/blob/master/impedancefitter/tests/test_readin.py where you test the new functionality. That would be very helpful so that it remains functional also in future versions.

And perfect that you refactor the code to be more coherent :)

**Comment by j-zimmermann** [comment]
Thanks for the tests and your contribution! 
I will merge the code now and then refactor.
I hope to be done in the next two weeks and I will publish a new release then (some dependencies need updates and some tests are broken because of outdated dependencies).
Always feel free to open an issue or a PR, I am happy to receive feedback to improve the code.

**Comment by Sukikui** [comment]
Perfect! Thanks to you
